### PR TITLE
Restart negotiations

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: https://github.com/golemfactory/yagna/issues/1128

### CODE REVIEW

**Review by jiivan — DISMISSED** [review]
LGTM

File: core/market/src/db/dao/proposal.rs [hunk 1]
```diff
@@ -72,6 +72,10 @@ impl<'c> ProposalDao<'c> {
                 return Err(SaveProposalError::AlreadyCountered(prev_proposal));
             }
 
+            // Proposal could be in Rejected state. It is ok to counter Rejected Proposal,
+            // but we must change state of prev Proposal to Draft.
+            update_proposal_state(conn, &prev_proposal, ProposalState::Draft)?;
```
Comment (tworec) [inline core/market/src/db/dao/proposal.rs:77]: this unconditional state change worries me

File: core/market/tests/test_negotiations.rs [hunk 2]
```diff
@@ -944,3 +944,84 @@ async fn test_proposal_events_last() {
...
+        ProposalState::Draft
+    );
+    assert_eq!(
+        req_market
+            .get_proposal(&negotiation.proposal_id.clone().translate(Owner::Requestor))
+            .await
+            .unwrap()
+            .body
+            .state,
+        ProposalState::Draft
```
Comment (tworec) [inline core/market/tests/test_negotiations.rs:1025]: I'd expect some finalized agreement as a conclusion of restarted negotation

File: core/market/src/db/dao/proposal.rs [hunk 3]
```diff
@@ -63,18 +63,30 @@ impl<'c> ProposalDao<'c> {
     pub async fn save_proposal(&self, proposal: &Proposal) -> Result<(), SaveProposalError> {
         let proposal = proposal.body.clone();
         do_with_transaction(self.pool, move |conn| {
-            let prev_proposal = proposal
+            let prev_proposal_id = proposal
```
Comment (tworec) [inline core/market/src/db/dao/proposal.rs:66]: 👍

File: core/market/src/db/dao/proposal.rs [hunk 4]
```diff
@@ -63,18 +63,30 @@ impl<'c> ProposalDao<'c> {
...
-            update_proposal_state(conn, &prev_proposal, ProposalState::Draft)?;
+            let prev_proposal: DbProposal = dsl::market_proposal
+                .filter(dsl::id.eq(&prev_proposal_id))
+                .first(conn)
+                .optional()?
+                .ok_or(SaveProposalError::NoPrevious(proposal.id.clone()))?;
+
+            // If previous Proposal was rejected, we must change it's state back.
+            if prev_proposal.state == ProposalState::Rejected {
+                // If rejected Proposal doesn't have prev_proposal_id, it was Initial Proposal.
```
Comment (tworec) [inline core/market/src/db/dao/proposal.rs:83]: 👍  great comments

File: core/market/tests/test_negotiations.rs [hunk 5]
```diff
@@ -984,7 +985,7 @@ async fn test_restart_negotiations() {
         .await
         .unwrap();
 
-    let prov_proposal_id = req_market
+    let req_proposal_id = req_market
```
Comment (tworec) [inline core/market/tests/test_negotiations.rs:988]:
```suggestion
    // now the magic begins. We are countering a rejected proposal
    let req_proposal_id = req_market
```

File: core/market/tests/test_negotiations.rs [hunk 6]
```diff
@@ -1026,6 +1030,45 @@ async fn test_restart_negotiations() {
             .state,
         ProposalState::Draft
     );
+
+    // Agents should be able to sign Agreement after rejection.
```
Comment (tworec) [inline core/market/tests/test_negotiations.rs:1034]: 👍

File: core/market/tests/test_negotiations.rs [hunk 7]
```diff
@@ -1123,3 +1166,75 @@ async fn test_negotiations_after_agreement_rejected() {
...
+                &initial_proposal
+                    .get_proposal_id()
+                    .unwrap()
+                    .translate(Owner::Requestor)
+            )
+            .await
+            .unwrap()
+            .body
+            .state,
+        ProposalState::Initial
```
Comment (tworec) [inline core/market/tests/test_negotiations.rs:1238]: 👍

**Review by tworec — APPROVED** [review]
great job!

**Pull request merged**